pkg/kube/ctl.go

```diff
@@ -1,24 +1,139 @@
 package kube
 
 import (
+	"context"
+	"io"
+
+	"github.com/cockroachdb/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/kubernetes/scheme"
 	"k8s.io/client-go/rest"
 	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/tools/remotecommand"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type (
+	Object    = client.Object
+	ObjectKey = client.ObjectKey
 )
 
+// FromEnv returns a [Ctl] for the default context in $KUBECONFIG.
 func FromEnv() (*Ctl, error) {
 	loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
 	configOverrides := &clientcmd.ConfigOverrides{}
 	kubeConfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, configOverrides)
-	kubeConfig.ClientConfig()
 	config, err := kubeConfig.ClientConfig()
-	return &Ctl{config: config}, err
+	if err != nil {
+		return nil, err
+	}
+
+	c, err := client.New(config, client.Options{})
+	if err != nil {
+		return nil, err
+	}
+
+	return &Ctl{
+		config: config,
+		client: c,
+	}, nil
 }
 
+// Ctl is a Kubernetes client inspired by the shape of the `kubectl` CLI with a
+// focus on being ergonomic.
 type Ctl struct {
 	config *rest.Config
+	client client.Client
 }
 
-// RestConfig returns a deep copy of the *[rest.Config] used by this [Ctl].
+// RestConfig returns a deep copy of the [rest.Config] used by this [Ctl].
 func (c *Ctl) RestConfig() *rest.Config {
 	return rest.CopyConfig(c.config)
 }
+
+// Get fetches the latest state of an object into `obj` from Kubernetes.
+// Usage:
+//
+//	var pod corev1.Pod
+//	ctl.Get(ctx, kube.ObjectKey{Namespace: "", Name:""}, &pod)
+func (c *Ctl) Get(ctx context.Context, key ObjectKey, obj Object) error {
+	if err := c.client.Get(ctx, key, obj); err != nil {
+		return errors.WithStack(err)
+	}
+	return nil
+}
+
+func (c *Ctl) Create(ctx context.Context, obj Object) error {
+	if err := c.client.Create(ctx, obj); err != nil {
+		return errors.WithStack(err)
+	}
+	return nil
+}
+
+func (c *Ctl) Delete(ctx context.Context, obj Object) error {
+	if err := c.client.Delete(ctx, obj); err != nil {
+		return errors.WithStack(err)
+	}
+	return nil
+}
+
+type ExecOptions struct {
+	Container string
+	Command   []string
+	Stdin     io.Reader
+	Stdout    io.Writer
+	Stderr    io.Writer
+}
+
+// Exec runs `kubectl exec` on the given Pod in the style of [exec.Command].
+func (c *Ctl) Exec(ctx context.Context, pod *corev1.Pod, opts ExecOptions) error {
+	if opts.Container == "" {
+		opts.Container = pod.Spec.Containers[0].Name
+	}
+
+	// Apparently, nothing in the k8s SDK, except exec'ing, uses RESTClientFor.
+	// RESTClientFor checks for GroupVersion and NegotiatedSerializer which are
+	// never set by the config loading tool chain.
+	// The .APIPath setting was a random shot in the dark that happened to work...
+	// Pulled from https://github.com/kubernetes/kubectl/blob/master/pkg/cmd/util/kubectl_match_version.go#L115
```

Comment:
NIT: Use permalink, so that the reference will not rotten.

```suggestion
	// Pulled from https://github.com/kubernetes/kubectl/blob/acf4a09f2daede8fdbf65514ade9426db0367ed3/pkg/cmd/util/kubectl_match_version.go#L115
```